src/flint/types/fmpz_mod_poly.pyx

```diff
@@ -0,0 +1,1297 @@
+        cdef fmpz_mod_poly res
+
+        # Case when left and right are already fmpz_mod_poly
+        if typecheck(left, fmpz_mod_poly) and typecheck(right, fmpz_mod_poly):
+            if not (<fmpz_mod_poly>left).ctx == (<fmpz_mod_poly>right).ctx:
+                raise ValueError("moduli must match")
+
+        # Case when right is not fmpz_mod_poly, try to convert to fmpz
+        elif typecheck(left, fmpz_mod_poly):
+            right = (<fmpz_mod_poly>left).ctx.any_as_fmpz_mod_poly(right)
+            if right is NotImplemented:
+                return NotImplemented
+
+        # Case when left is not fmpz_mod_poly, try to convert to fmpz
+        else:
+            left = (<fmpz_mod_poly>right).ctx.any_as_fmpz_mod_poly(left)
+            if left is NotImplemented:
+                return NotImplemented
+
+        if not right.leading_coefficient().is_unit():
+            raise ValueError(f"The leading term of {right} must be a unit modulo N")
+
+        res = fmpz_mod_poly.__new__(fmpz_mod_poly)
+        res.ctx = (<fmpz_mod_poly>left).ctx
+        fmpz_mod_poly_div(
+            res.val, (<fmpz_mod_poly>left).val, (<fmpz_mod_poly>right).val, res.ctx.mod.val
+        )
+        return res 
+
+    def __floordiv__(self, other):
+        return fmpz_mod_poly._floordiv_(self, other)
+
+    def __rfloordiv__(self, other):
+        return fmpz_mod_poly._floordiv_(other, self)
+
+    def __pow__(self, e):
+        cdef fmpz_mod_poly res
+        if e < 0:
+            raise ValueError("Exponent must be non-negative")
+
+        res = fmpz_mod_poly.__new__(fmpz_mod_poly)
+        res.ctx = self.ctx
+        fmpz_mod_poly_pow(
+            res.val, self.val, (<ulong>e), self.ctx.mod.val
+        )
+        return res
+
+    def shift(self, slong n):
+        """
+        Returns ``self`` shifted by ``n`` coefficients. If ``n`` is positive,
+        zero coefficients are inserted, when ``n`` is negative, if ``n`` is 
+        greater than or equal to the length of ``self``, the zero polynomial
+        is returned.
+
+            >>> R = fmpz_mod_poly_ctx(163)
+            >>> f = R([1,2,3])
+            >>> f.shift(0)
+            3*x^2 + 2*x + 1
+            >>> f.shift(1)
+            3*x^3 + 2*x^2 + x
+            >>> f.shift(4)
+            3*x^6 + 2*x^5 + x^4
+            >>> f.shift(-1)
+            3*x + 2
+            >>> f.shift(-4)
+            0
+
+        """
+        if n == 0:
+            return self
+
+        cdef fmpz_mod_poly res
+        res = fmpz_mod_poly.__new__(fmpz_mod_poly)
+        res.ctx = self.ctx
+
+        if n > 0:
+            fmpz_mod_poly_shift_left(
+                res.val, self.val, n, self.ctx.mod.val
+            )
+        elif n < 0:
+            fmpz_mod_poly_shift_right(
+                res.val, self.val, -n, self.ctx.mod.val
+            )
+
+        return res          
+
+    def __lshift__(self, n):
+        return self.shift(n)
+
+    def __rshift__(self, n):
+        return self.shift(-n)
```

Comment:
I'm not sure about this. Do any of the other poly types have this?
```
In [16]: f
Out[16]: 3*x^2 + 2*x + 1

In [17]: f << 2
Out[17]: 3*x^4 + 2*x^3 + x^2

In [18]: f >> 2
Out[18]: 3
```
Is this a common operation or are there other environments that define the operator in this way?

Reply:
Is that particularly useful though?

A decision should be made for all poly types if we are to add methods like this: they all need to be consistent.

Reply:
```py
In [21]: R = fmpz_mod_poly_ctx(2**255 - 19)

In [22]: f = R.random_element(degree=1000)

In [23]: g = x**100

In [24]: f // g == f >> 100
Out[24]: True

In [25]: %timeit f >> 100
139 µs ± 6.5 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [26]: %timeit f // g
384 µs ± 28 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [27]: f * g == f << 100
Out[27]: True

In [28]: %timeit f*g
477 µs ± 9.6 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [29]: %timeit f << 100
158 µs ± 1.81 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

I believe it is useful, and have used it myself in certain algorithms.

Reply:
I can see that the operation is useful but the comparison should be
```python
In [20]: %timeit f << 100
279 µs ± 17.9 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [21]: %timeit f.shift(100)
275 µs ± 11.3 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [22]: f << 100 == f.shift(100)
Out[22]: True
```
The question is: if we already have `.shift()` then do we want to add `<<` as well?

Something I am less keen on for the name "shift" is that it is not completely obvious to me whether it multiplies or divides. I would probably call this something like `mul_xn` and `div_xn` or something just to make it clear which way it goes.

Using operators is good for very common operations where the meaning is intuitive and the precedence etc rules for the operator make sense for the operation. If the operation is not that common or if it is not completely obvious what the operation does then using operators can just make things confusing.

Here is an example of something that might seem unexpected with this operator:
```python
In [24]: z = fmpz(5)

In [25]: z << 2
Out[25]: 20

In [26]: R(z) << 2
Out[26]: 5*x^2
```
As far as I am aware every other operator here is defined in such a way that if it is defined for both types then it would do equivalent things mapping homorphically from one domain to another:
```python
In [28]: +R(z) == R(+z)
Out[28]: True

In [29]: -R(z) == R(-z)
Out[29]: True

In [30]: R(z+1) == R(z)+1
Out[30]: True

In [31]: R(z)/2 == R(z/2) # not defined for both
---------------------------------------------------------------------------
TypeError 

In [32]: R(z**2) == R(z**2)
Out[32]: True

In [33]: R(z << 2) == R(z) << 2  # Not equivalent
Out[33]: False
```

Reply:
Ahh ok, again here I was just inpired by the sagemath behaviour, as well as the names of the flint functions, they're defined as the left and right shifts of a polynomial so `__rshift__` and `__lshift__` seemed natural. 

Happy to modify if you want though

Reply:
The Flint functions are called `shift_left` and `shift_right`. There is obvious value in matching the names Flint uses because this is a thin wrapper around Flint and so it is useful to know which method here corresponds to which Flint function. Those names are also better than `shift` because they indicate a direction.

There is also some value in being compatible with Sage for something like `p << 5` but it is worth remembering that probably many of Sage's API design choices are not good ones. I don't mean that Sage is badly designed but just that systems like Sage and SymPy will have a lot of legacy code that got added years ago by *someone* who thought that something looked nice at the time. Realistically if you were designing things from scratch then I expect that much of the API in these systems would just be removed. Excessive operator overloading is a good example of something that seems clever at the time until you realise that the operators are not really coherent in the end.

Does there exist any system besides Sage (e.g. Pari etc) in which left and right shift operators are defined for polynomials?

Reply:
> systems like Sage and SymPy will have a lot of legacy code that got added years ago by _someone_ who thought that something looked nice at the time. Realistically if you were designing things from scratch then I expect that much of the API in these systems would just be removed

The reason these things don't get removed is precisely because it is much easier to add a poorly thought out public interface than it is to remove it later once code is written that depends on it. In this situation of adding lots of public interface methods you should think carefully about whether those interfaces make a good design.